crates/kv-store/src/block.rs

```diff
@@ -0,0 +1,659 @@
+        if is_large{
+            return LargeValueBlock::decode(raw_block_and_check, key, compression_type).map(Block::Large).unwrap()
+        }
+        NormalBlock::decode(raw_block_and_check, key, compression_type).map(Block::Normal).unwrap()
+    }
+
+    pub fn len(&self)->usize{
+        match self{
+            Block::Normal(block)=>block.offsets.len(),
+            Block::Large(_)=>1,
+        }
+    }
+
+    pub fn is_empty(&self)->bool{
+        match self{
+            Block::Normal(block)=>block.offsets.is_empty(),
+            Block::Large(_)=>false,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct BlockBuilder {
+    data: Vec<u8>,
+    offsets: Vec<u16>,
+    block_size: usize,
+    // for key compression
+    first_key: Bytes,
+    is_large: bool,
+}
+
+impl BlockBuilder {
+    pub fn new(block_size: usize) -> Self {
+        Self {
+            data: Vec::new(),
+            offsets: Vec::new(),
+            block_size,
+            first_key: Bytes::new(),
+            is_large:false
+        }
+    }
+
+    fn estimated_size(&self) -> usize {
+        if self.is_large{
+            self.data.len()
+        }else{
+            // key-value pairs number
+            SIZE_OF_U16 +
+            // offsets 
+            self.offsets.len() * SIZE_OF_U16 + 
+            // key-value pairs data
+            self.data.len() +
+            // checksum
+            SIZE_OF_U32
+        }
+    }
+
+    pub fn is_empty(&self)->bool{
+        !self.is_large && self.offsets.is_empty()
+    }
+
+    /// Add a key-value pair to the block.
+    /// Returns true if the key-value pair is added successfully, false the block is full.
+    /// 
+    /// ┌─────────────────────────────────────────────────────┐
+    /// │  Key Value Chunk                                    │
+    /// │┌ ─ ─ ─ ─ ─ ─ ─ ─ ┬ ─ ─ ─ ─ ─ ─ ─┌ ─ ─ ─ ─ ─┬ ─ ─ ─ ┐│
+    /// │ common prefix len key suffix len│key suffix│ value ││
+    /// ││       u8        │     u16      │  bytes   │ bytes ││
+    /// │ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ┘ ─ ─ ─ ┘│
+    /// └─────────────────────────────────────────────────────┘
+    /// 
+    pub fn add(&mut self, key: &[u8], value: &[u8]) -> bool {
+        debug_assert!(!key.is_empty(), "key cannot be empty");
+        if  self.first_key.is_empty() && value.len() > self.block_size {
+            self.data.put(value);
+            self.is_large = true;
+            self.first_key = Bytes::copy_from_slice(key);
+            return true;
+        }
+
+        // whether the block is full
+        if self.estimated_size() + key.len() + value.len() + SIZE_OF_U8 + SIZE_OF_U16 > self.block_size && !self.first_key.is_empty() {
+            return false;
+        }
+
+        if self.first_key.is_empty() {
+            self.first_key = Bytes::copy_from_slice(key);
+        }
+        self.offsets.push(self.data.len() as u16);
+        let (common, suffix) = get_common_prefix_len_and_strip(key, &self.first_key);
```

Comment:
So is there a special behavior for the first chunk or not? No?

Reply:
What kind of special behavior? The first one needs not to encode the key ?

Reply:
Yep